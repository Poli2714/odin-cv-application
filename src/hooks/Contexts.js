import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
