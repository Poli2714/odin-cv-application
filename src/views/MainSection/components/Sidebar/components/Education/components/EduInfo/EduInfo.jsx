import { useContext } from 'react';
import PropTypes from 'prop-types';

import { CharacterLimit } from 'src/components/ui';
import { InputWrapper } from 'src/components/layout';
import { Label, TextArea } from 'src/components/forms';
import { EducationContext } from 'src/hooks/Contexts';

const limit = 100;
const rows = 5;

function EduInfo({ id }) {
  const { educations, setEducations } = useContext(EducationContext);
  const { eduInfo } = educations[id];

  const handleChangeEduInfo = e => {
    if (e.target.value.length > limit) return;
    setEducations({
      ...educations,
      [id]: { ...educations[id], eduInfo: e.target.value },
    });
  };

  return (
    <InputWrapper>
      <Label labelName="Additional Information" labelFor="eduInfo" />
      <CharacterLimit limit={limit} value={eduInfo}>
        <TextArea
          name="eduInfo"
          maxLength={limit}
          rows={rows}
          value={eduInfo}
          onChange={handleChangeEduInfo}
        />
      </CharacterLimit>
    </InputWrapper>
  );
}

EduInfo.propTypes = {
  id: PropTypes.string,
};

export default EduInfo;
