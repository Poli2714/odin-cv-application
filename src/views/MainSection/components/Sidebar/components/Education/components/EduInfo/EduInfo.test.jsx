import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import EduInfo from './EduInfo';
import { EducationContext } from 'src/hooks/Contexts';

vi.mock('src/components/ui', () => ({
  CharacterLimit: vi.fn(({ children, value }) => (
    <div>
      <span data-testid="limit">{value.length} / 100</span>
      {children}
    </div>
  )),
}));

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="eduInfo">Mock label</label>),
  TextArea: vi.fn(({ onChange, value }) => (
    <textarea
      id="eduInfo"
      maxLength={100}
      onChange={onChange}
      rows={5}
      value={value}
    ></textarea>
  )),
}));

function MockParentComponent({ id }) {
  const [educations, setEducations] = useState({
    fakeID: { eduInfo: '' },
    fakeID2: { eduInfo: 'Test info' },
  });

  return (
    <EducationContext.Provider value={{ educations, setEducations }}>
      <EduInfo id={id} />
    </EducationContext.Provider>
  );
}

test('renders EduInfo', () => {
  render(<MockParentComponent id="fakeID" />);

  expect(screen.getByTestId('input-wrapper')).toMatchSnapshot();
  expect(screen.getByLabelText('Mock label')).toHaveValue('');
});

test('renders EduInfo with different initial value', () => {
  render(<MockParentComponent id="fakeID2" />);

  expect(screen.getByLabelText('Mock label')).toHaveValue('Test info');
});

test('updates eduInfo value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="fakeID" />);
  const textarea = screen.getByLabelText('Mock label');
  const limitSpan = screen.getByTestId('limit');

  await user.type(textarea, 'Testing component');
  expect(textarea).toHaveValue('Testing component');
  expect(limitSpan).toHaveTextContent('17 / 100');

  await user.type(textarea, '. Test!');
  expect(textarea).toHaveValue('Testing component. Test!');
  expect(limitSpan).toHaveTextContent('24 / 100');
});

test("doesn't update eduInfo value if a user types over character limit", async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="fakeID" />);
  const textarea = screen.getByLabelText('Mock label');
  const limitSpan = screen.getByTestId('limit');

  await user.type(
    textarea,
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum eget justo vel placerat feugiat.',
  );
  expect(textarea.value).toHaveLength(100);
  expect(limitSpan).toHaveTextContent('100 / 100');

  await user.type(textarea, ' Hey!');
  expect(textarea.value).toHaveLength(100);
  expect(limitSpan).toHaveTextContent('100 / 100');
});
