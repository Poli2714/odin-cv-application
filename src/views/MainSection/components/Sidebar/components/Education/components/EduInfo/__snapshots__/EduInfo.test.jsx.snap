// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders EduInfo 1`] = `
<div
  data-testid="input-wrapper"
>
  <label
    for="eduInfo"
  >
    Mock label
  </label>
  <div>
    <span
      data-testid="limit"
    >
      0
       / 100
    </span>
    <textarea
      id="eduInfo"
      maxlength="100"
      rows="5"
    />
  </div>
</div>
`;
